File: elixir/secret-handshake/lib/secret_handshake.ex
defmodule SecretHandshake do
  @doc """
  Determine the actions of a secret handshake based on the binary
  representation of the given `code`.

  If the following bits are set, include the corresponding action in your list
  of commands, in order from lowest to highest.

  1 = wink
  10 = double blink
  100 = close your eyes
  1000 = jump

  10000 = Reverse the order of the operations in the secret handshake
  """
  @spec commands(code :: integer) :: list(String.t())
  def commands(code) do
    []
    |> check_for_bit(Bitwise.band(code, 0b00001))
    |> check_for_bit(Bitwise.band(code, 0b00010))
    |> check_for_bit(Bitwise.band(code, 0b00100))
    |> check_for_bit(Bitwise.band(code, 0b01000))
    |> check_for_bit(Bitwise.band(code, 0b10000))
  end

  def check_for_bit(actions, 0b00001), do: actions ++ ["wink"]
  def check_for_bit(actions, 0b00010), do: actions ++ ["double blink"]
  def check_for_bit(actions, 0b00100), do: actions ++ ["close your eyes"]
  def check_for_bit(actions, 0b01000), do: actions ++ ["jump"]
  def check_for_bit(actions, 0b10000), do: Enum.reverse(actions)
  def check_for_bit(actions, _), do: actions
end

File: go/secret-handshake/secret_handshake.go
package secret

var actions = map[uint]string{
	0b0001: "wink",
	0b0010: "double blink",
	0b0100: "close your eyes",
	0b1000: "jump",
}

func Handshake(code uint) []string {
	var secretHandshake []string
	for i := 0; i < len(actions); i++ {
		var bitIndex uint = 0x1 << i
		if code&bitIndex > 0 {
			secretHandshake = append(secretHandshake, actions[bitIndex])
		}
	}

	if shouldReverse(code) {
		secretHandshake = reverse(secretHandshake)
	}

	return secretHandshake
}

func shouldReverse(code uint) bool {
	var reverseBit = len(actions) + 1
	return code&(1<<uint(reverseBit-1)) != 0
}

func reverse(actions []string) []string {
	for i, j := 0, len(actions)-1; i < j; i, j = i+1, j-1 {
		actions[i], actions[j] = actions[j], actions[i]
	}

	return actions
}